File: src/main/java/com/github/zxh/classpy/classfile/constant/ConstantFieldrefInfo.java
package com.github.zxh.classpy.classfile.constant;

import com.github.zxh.classpy.classfile.datatype.U2;

/*
CONSTANT_Fieldref_info {
    u1 tag;
    u2 class_index;
    u2 name_and_type_index;
}
*/
public class ConstantFieldrefInfo extends ConstantInfo {

    {
        u2("class_index");
        u2("name_and_type_index");
    }
    
    @Override
    protected String loadDesc(ConstantPool pool) {
        int classIndex = super.getUInt("class_index");
        int nameAndTypeIndex = super.getUInt("name_and_type_index");

        ConstantClassInfo classInfo = pool.getClassInfo(classIndex);
        String className = pool.getUtf8String(classInfo.getNameIndex());
        ConstantNameAndTypeInfo nameAndTypeInfo = pool.getNameAndTypeInfo(nameAndTypeIndex);
        String fieldName = pool.getUtf8String(nameAndTypeInfo.getNameIndex());
        return className + "." + fieldName;
    }
    
}

File: src/main/java/com/github/zxh/classpy/classfile/constant/ConstantFieldrefInfo.kt
package com.github.zxh.classpy.classfile.constant;

/*
CONSTANT_Fieldref_info {
    u1 tag;
    u2 class_index;
    u2 name_and_type_index;
}
*/
class ConstantFieldrefInfo : ConstantInfo() {

    init {
        u2("class_index");
        u2("name_and_type_index");
    }
    
    override fun loadDesc(cp: ConstantPool): String {
        val classIndex = super.getInt("class_index");
        val nameAndTypeIndex = super.getInt("name_and_type_index");

        val classInfo = cp.getClassInfo(classIndex);
        val className = cp.getUtf8String(classInfo.getNameIndex());
        val nameAndTypeInfo = cp.getNameAndTypeInfo(nameAndTypeIndex);
        val fieldName = cp.getUtf8String(nameAndTypeInfo.getNameIndex());
        return className + "." + fieldName;
    }
    
}
